README.md
# word-conversion

A Clojure library to convert numerical representation of whole numbers to British English representation, using correct basic grammar.


## Usage

As this is a library and does not include a user interface.  Open a REPL and load the namespace then evaluate the examples in the design journal section.

### Command Line / Terminal
In a terminal window, change to the root of the `word-conversion` project and run the command `lein repl`.  Once the repl starts either include the project namespace in the current `user` namespace, using `(require 'word-conversion.core)`

### Clojure aware editors
Or open this project in your favourite editor, run a repl and switch to the `word-conversion.core` namespace.

## License

Creative Commons Attribution Share-Alike 4.0 International

Copyright © 2019 John Stevenson


## Number to Words Conversion

### Problem Description
Create a Clojure library, suitable for use in a server-side application,that can take a Java int in the range 0 to 999,999,999 inclusive andreturns the equivalent number, as a String, in British English words.

### Sample Data

|     Input | Output                                                                                                   |
|-----------|----------------------------------------------------------------------------------------------------------|
|         0 | zero                                                                                                     |
|         1 | one                                                                                                      |
|        21 | twenty  one                                                                                              |
|       105 | one hundred  and five                                                                                    |
|       123 | one hundred and twenty three                                                                             |
|      1005 | one thousand and five                                                                                    |
|      1042 | one thousand and forty two                                                                               |
|      1105 | one thousand one hundred and five                                                                        |
|  56945781 | fifty six million nine hundred and forty five thousand seven hundred and eighty one                      |
| 999999999 | nine hundred and ninety  nine million nine hundred and ninety nine thousand nine hundred and ninety nine |


### Guidelines

●The solution must be correct.
Please pay attention to the specific conventions of British English, particularly concerning the use of ‘and’between certain phrases.

●The solution is not expected to involve a command line or GUI application
– we’re looking for a Clojure library that could be packaged as a jar andused in a larger application.


## Analysis of the problem statement

### Notes
* the British English words are all lower case, no capitalisation is required for the start of each sentence.


### Initial thoughts for a solution

Taking a simplest set first approach:

#### Convert each word to a string using a simple lookup

Define a dictionary that maps numbers 0 to 9 to their word equivalents.


#### Insert the number scales in the correct places

Insert `hundred`, `thousand`, and `million` number scale names into the words.

> Wondering if this needs to be done first or at least at the same time as converting from numbers to strings.


#### Insert the correct grammar into the words (and)

Parse a collection of words inserting `and` at the appropriate point in the sentence.  The rules seem to be positional, so we can just process the word strings for the correct places to insert `and`.

The rules seem to be:
- numbers between 101 and 999 have and after the first number (or before the last two numbers).
- the above rule should be applied all at each number scale for numbers larger than 1001


### Additional thoughts ###

#### Splitting up a number into digits ####

How can we split up a number into its digits?  Unlike string, a number is not seen as a collection.  We can convert each number to a string, this would allow us to split the number into individual digits, although if using clojure.core functions then the string will be treated as a collection of characters.  Can we convert characters back to numbers (should we need to)?  Strings can be converted back to numbers easily with the `java.lang.Integer` class.

If we use strings or characters in our dictionary lookup, then we don't need to convert back.

Some experiments in partitioning are in the design journal.


#### replace each number with its whole value ####

If we took each digit in the original number and converted it to its representative number in terms of position, then the dictionary lookup becomes much simpler.

For example, if the original number is 12345, then we would first generate a sequence of `[10000 2000 300 40 5]`.

This approach seems to be pretty obvious now I thought about it.


## Interesting functions to consider

`partition` will group consistently and provide an simple way to insert new strings into the right parts of the words.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-04-01
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-04-01
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/word-conversion/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/word-conversion/compare/0.1.0...0.1.1


test/word_conversion/core_test.clj
(ns word-conversion.core-test
  (:require [clojure.test :refer :all]
            [word-conversion.core :as sut]
            [word-conversion.dictionaries :refer [british-english-dictionary]]))



(deftest number->british-english-test
  "Test the conversion of numerical whole numbers to
   strings containing British English words as a sentence"

  (testing "Sample data tests - single values"
    (is (= "zero" (sut/digit->word 0 sut/british-english-numbers)))
    (is (= "one"  (sut/digit->word 1 sut/british-english-numbers))))

  (testing "Sample data tests - double figure values"
    (is (= "20" (sut/positional-number-string "21"))))

  (testing "Generating sequences of numbers with positional number level"
    (is (= ["20" "1"]   (sut/number-sequence 21)))
    (is (= ["2000"]     (sut/number-sequence 2000)))
    (is (= ["2000" "1"] (sut/number-sequence 2001))))

  (testing "Convert sequences of numbers to words"
    (is (= ["zero"]
           (sut/word-sequence british-english-dictionary ["0"])))
    (is (= ["one"]
           (sut/word-sequence british-english-dictionary ["1"])))
    (is (= ["twenty"]
           (sut/word-sequence british-english-dictionary ["20"])))
    (is (= ["twenty" "one"]
           (sut/word-sequence british-english-dictionary ["20" "1"])))
    (is (= ["ninety" "nine"]
           (sut/word-sequence british-english-dictionary ["90" "9"])))
    (is (= ["one hundred"]
           (sut/word-sequence british-english-dictionary ["100"])))
    (is (= ["one hundred" "one"]
           (sut/word-sequence british-english-dictionary ["100" "1"])))
    (is (= ["nine hundred" "ninety" "nine"]
           (sut/word-sequence british-english-dictionary ["900" "90" "9"])))
    (is (= ["one thousand"]
           (sut/word-sequence british-english-dictionary ["1000"])))
    (is (= ["one thousand" "one"]
           (sut/word-sequence british-english-dictionary ["1000" "1"])))))



src/word_conversion/dictionaries.clj
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Dictionaries for Number to Word Conversion
;;
;; Dictionaries used as lookup tables to convert numbers into their
;; respective word representations
;;
;; Author(s): John Stevenson
;; Date created: 7th April 2019
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(ns word-conversion.dictionaries)

;; Specific number lookups
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; zero to nine
(def single-digit
  {"0" "zero"
   "1" "one"
   "2" "two"
   "3" "three"
   "4" "four"
   "5" "five"
   "6" "six"
   "7" "seven"
   "8" "eight"
   "9" "nine"})

(def teens
  {"11" "eleven"
   "12" "twelve"
   "13" "thirteen"
   "14" "fourteen"
   "15" "fifteen"
   "16" "sixteen"
   "17" "seventeen"
   "18" "eighteen"
   "19" "nineteen"})

;; tens
(def tens
  {"10" "ten"
   "20" "twenty"
   "30" "thirty"
   "40" "forty"
   "50" "fifty"
   "60" "sixty"
   "70" "seventy"
   "80" "eighty"
   "90" "ninety"})


(def british-english-dictionary
  "Combination of dictionaries that include British English lookup
  values"
  (merge single-digit teens tens))


;; Generic Number levels
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; A dictionary of number levels
;; Indexed by the size of the number as a string
;; 100 has 3 characters, 1000000 has 7 characters, etc.

;; hundreds, thousands and millions all the same term regardless
;; of prefix number (one thousand, two thousand, etc.)
;; simply add number level as post-fix

(def number-levels
  "List of number levels."
  {3 "hundred" 4 "thousand" 6 "hundred thousand" 7 "million" 10 "billion" 13 "trillion"})


src/word_conversion/core.clj
(ns word-conversion.core
  (:require [word-conversion.dictionaries :refer [british-english-dictionary
                                                  number-levels]]))

(def british-english-numbers
  {0 "zero"
   1 "one"})

(defn digit->word
  "Converts a numeric whole number (digit) into a word representation
  Returns: java.lang.string"

  [digit dictionary]

  (get dictionary digit))


(defn positional-number-string
  "Round a number down to its positional number level.

  Examples: 2345 becomes 2000"
  [number-string]
  (apply str
         (cons (first number-string)
               (map (constantly "0") (rest number-string)))))


(defn clean-number-sequence
  "All number strings that are pronounced start with something other than zero.
  All zero numbers are removed, except where zero is the only value in the sequence.

  Arguments: Vector of positional numbers as strings
  Return: Vector of positional numbers as strings"

  [number-string]

  (if (= ["0"] number-string)
    ["0"]
    (filter #(not= \0 (first %)) number-string)))


(defn number-sequence
  "Convert a number into a sequence of numbers that also represent the
    number level

  Arguments: Integer or Long
  Return: Vector of strings"

  [number]

  {:pre [(<= 0 number)]} ; Exception if function called with negative number

  (loop [current-string      (str number)
         sequence-of-numbers []]
    (if (empty? current-string)
      (clean-number-sequence sequence-of-numbers)
      (recur (rest current-string)
             (conj sequence-of-numbers
                   (positional-number-string current-string))))))




(defn multiple-number-levels
  [number]
  (let [remainder (rem (count number) 3)
        levels    (quot (count number) 3)]
    (if (> 1 levels)
      (apply str (take remainder number) (multiple-number-levels (drop remainder)))
      number)))


(defn get-prefix-number
  "Gets the prefix of a number without the number level.

  Number levels go up in groups of three, so 1000 returns 1, 10000 returns 10.

  Arguments: String (representing a number with its positional number level)
  Return: String (representing just the prefix value)"

  [number-string]
  (let [prefix (rem (count number-string) 3)
        levels (quot (count number-string) 3)]

    (cond
      (= prefix 0) (str (first number-string))
      (> 1 levels) (multiple-number-levels number-string)
      :else        (apply str (take prefix number-string)))))

(get-prefix-number "1000")
;; => "1"
(get-prefix-number "10000")
;; => "10"
(get-prefix-number "100000")
;; => "1"

;; if
;; 100 = 1
;; 1000 = 1
;; 10,000 = 10
;; 100,000 = 100
;; 1,000,000 = 1
(defn word-sequence
  "Convert numbers in a sequence to their corresponding words, using a dictionary

  Arguments: dictionary lookup table (hash-map), sequence of integer/long numbers
  Return: sequence of words (vector of strings)"
  [dictionary number-sequence]

  (map (fn [positional-number]
         (if (<= (count positional-number) 3)
           (get dictionary positional-number)
           (str (get dictionary (get-prefix-number positional-number))
                " "
                (get number-levels (count positional-number)))))
       number-sequence))




;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; REPL design journal
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; Convert numbers to words using a hash-map dictionary
;; convert word using dictionary
#_(digit->word 0 british-english-numbers)


;; Splitting up a number into individual parts
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; We cant partition a number so lets change it to a string.  A string is a collection of chars after all.

;; convert to a string
#_(str 24)
;; => "24"

#_(partition 1 "1234")
;; => ((\1) (\2) (\3) (\4))


;; Using for just splits to individual words
#_(defn partition-number
    [numbers]
    (for [number numbers]
      [(Integer. (str number))]))

#_(partition-number "1234")
;; => ([1] [2] [3] [4])


;; create a sequence of strings representing numbers
#_(clojure.string/split "24" #"")
;; => ["2" "4"]

;; We loose some of the sense of number level though.
;; Is 2 the value two or is it twenty?
;; We would need to track where we are in the sequence, relative to the end

;; recursive function to create a sequence of positional numbers
;;;;;;;;;;;;;;;;;;;;;;;;
;; get the first value
;; map (constantly 0) over the rest
;; conjoin
;; recur...

#_(defn positional-numbers-recursive
    [numbers]
    (let [string-numbers (str numbers)]
      (conj
        (map (constantly 0) (rest string-numbers))
        (first string-numbers))))

#_(positional-numbers-recursive 21)
;; => (\2 0)

;; still not quite the right format

#_(conj (first "2345") (map (constantly "0") (rest "2345")))


#_(map (constantly "0") (rest "2345"))
;; => ("0" "0" "0")



#_(mapcat (constantly "0") (rest "2345"))
;; => (\0 \0 \0)


#_(merge [] (first "2345") (map (constantly "0") (rest "2345")))
;; => [\2 ("0" "0" "0")]


#_(cons (first "2345") (map (constantly "0") (rest "2345")))
;; => (\2 "0" "0" "0")

#_(apply str
         (cons (first "2345") (map (constantly "0") (rest "2345"))))
;; => "2000"

#_(defn rounded-number-string
    "Round a number down to its number level.

  Examples: 2345 becomes 2000"
    [number-string]
    (apply str
           (cons (first number-string)
                 (map (constantly "0") (rest number-string)))))

#_(defn number-sequence
    "Convert a number into a sequence of numbers that also represent the
  number level"
    [number]
    {:pre [(<= 0 number)]}
    (loop [current-string      (str number)
           sequence-of-numbers []]
      (if (empty? current-string)
        sequence-of-numbers
        (recur (rest current-string)
               (conj sequence-of-numbers
                     (rounded-number-string current-string))))))

#_(number-sequence 23456)
;; => ["20000" "3000" "400" "50" "6"]

;; If called with a negative number, the function should return an error

#_(number-sequence -1)
;; java.lang.AssertionError
;; Assert failed: (<= 0 number)

;; Limitation
;; there is a bit of a gotcha in that if there is a zero value part way through a number,
;; then we would get 00 000 0000 etc.

;; We can just filter out all numbers that start with a zero,
;; as they would not be pronounced
;; would need to add a check for a single number in a sequence that is zero

#_(defn clean-number-sequence
    "All number strings should start with something other than zero"
    [number-string]
    (filter #(not= \0 (first %)) number-string))
;; => #'word-conversion.core/clean-number-sequence

#_(number-sequence 1024)
;; => ["1000" "000" "20" "4"]

#_(clean-number-sequence (number-sequence 1024))
;; => ("1000" "20" "4")


;; Add some defensive coding for a single zero sequence
#_(defn clean-number-sequence
    "All number strings should start with something other than zero"
    [number-string]
    (if (= ["0"] number-string)
      ["0"]
      (filter #(not= \0 (first %)) number-string)))
;; => #'word-conversion.core/clean-number-sequence

#_(clean-number-sequence ["0"])
;; => ["0"]


;; defining dictionaries
;;;;;;;;;;;;;

;; zero to nine
#_(def single-digit
    {"0" "zero"
     "1" "one"
     "2" "two"
     "3" "three"
     "4" "four"
     "5" "five"
     "6" "six"
     "7" "seven"
     "8" "eight"
     "9" "nine"})

#_(def teens
    {"11" "eleven"
     "12" "twelve"
     "13" "thirteen"
     "14" "fourteen"
     "15" "fifteen"
     "16" "sixteen"
     "17" "seventeen"
     "18" "eighteen"
     "19" "nineteen"})

;; tens
#_(def tens
    {"10" "ten"
     "20" "twenty"
     "30" "thirty"
     "40" "forty"
     "50" "fifty"
     "60" "sixty"
     "70" "seventy"
     "80" "eighty"
     "90" "ninety"})

;; hundreds, thousands and millions - no difference in numbers,
;; simply add level as postfix

;; create a sequential list of number levels
#_(def number-levels
    "List of number levels."
    ["hundred" "thousand" "hundred thousand" "million" "billion" "trillion"])


;; Alternative idea:
;; seems like many more combinations will need to be defined within dictionaries.
#_(def number-levels
    {2 ["two" "twenty" "two hundred" "two thousand" "two hundred thousand" "two million"]})

#_(def number-word-dictionary
    (merge single-digit teens tens))


#_(map #(get number-word-dictionary %) (word-sequence 42))
;; => ("forty" "two")


#_(map #(get number-word-dictionary %) (word-sequence 2))
;; => ("two")

;; numbers larger than 100
;;;;;;;;;;;;;;;;;;;;;;;;

;; If a number is larger than a hundred, we can just use the first digit
;; as a word and post-fix with the relevant number level (hundred,
;; thousand, etc.)

#_(word-sequence 101)
;; => ["100" "00" "1"]

#_(clean-word-sequence
    (word-sequence 101))
;; => ("100" "1")

;; If we count the number of digits in a stringified number,
;; then we get a consistent number level.

#_(count "200")
;; => 3

;; we can define a dictionary that is a lookup for number levels,
;; based on the string size

#_(def number-levels
    "List of number levels."
    {3 "hundred" 4 "thousand" 6 "hundred thousand" 7 "million" 10 "billion" 13 "trillion"})


(map (fn [positional-number]
       (if (<= (count positional-number) 2)
         (get british-english-dictionary positional-number)
         (str (get british-english-dictionary (str (first positional-number)))
              " "
              (get number-levels (count positional-number)))))
     ["100"])


(map (fn [positional-number]
       (if (<= (count positional-number) 2)
         (get british-english-dictionary positional-number)
         (str (get british-english-dictionary (str (first positional-number)))
              " "
              (get number-levels (count positional-number)))))
     ["100000"])
;; => ("one hundred thousand")


;; solution so far does not work for 10,000
;; so we need to add more logic to capture this case
(map (fn [positional-number]
       (if (<= (count positional-number) 2)
         (get british-english-dictionary positional-number)
         (str (get british-english-dictionary (str (first positional-number)))
              " "
              (get number-levels (count positional-number)))))
     ["10000"])
;; => ("one ")


(map (fn [positional-number]
       (if (<= (count positional-number) 2)
         (get british-english-dictionary positional-number)
         (str (get british-english-dictionary (str (first positional-number)))
              " "
              (get number-levels (count positional-number)))))
     ["10000"])


;; get the number level
;; 10000 => thousand

;; drop those numbers that represent the number level
;; 10000 => 10

(apply str
       (take (rem (count "10000") 3) "10000"))
;; => "10"

;; convert number after drop
;; 10 => "ten"

(defn get-prefix-number
  "Gets the prefix of a number without the number level.

  Number levels go up in groups of three, so "
  [number-string]
  (apply str
         (take (rem (count number-string) 3) number-string)))

(get-prefix-number "1")
;; => "1"
(get-prefix-number "10")
;; => "10"
(get-prefix-number "100")
;; => ""
(get-prefix-number "1000")
;; => "1"
(get-prefix-number "10000")
;; => "10"
(get-prefix-number "100000")
;; => ""
(get-prefix-number "1000000")
;; => "1"
(get-prefix-number "10000000")
;; => "10"
(get-prefix-number "100000000")
;; => ""

;; Partitioning
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(clojure.string/reverse "1234567")
;; => "7654321"

(partition-all
  3
  (clojure.string/reverse "1234567"))
;; => ((\7 \6 \5) (\4 \3 \2) (\1))

(defn partition-string-number
  [number-string]
  (partition-all
    3
    (clojure.string/reverse number-string)))

(partition-string-number "12001")
;; => ((\1 \0 \0) (\2 \1))


#_(partition-all 3 (word-sequence 1042))
;; => (("1000" "000" "40") ("2"))


;; seems like we need to reverse the sequence to partition correctly

#_(reverse
(word-sequence 1042))
;; => ("2" "40" "000" "1000")

#_(partition-all 3
(reverse
  (word-sequence 1042)))
;; => (("2" "40" "000") ("1000"))

#_(reverse
(partition-all 3
               (reverse
                 (word-sequence 1042))))
;; => (("1000") ("2" "40" "000"))

;; a double reverse seems ugly and its also reversing the elements
;; in each partition



#_(count (word-sequence 1042))
;; => 4

#_(mod
(count (word-sequence 1042))
3)
;; => 1


#_(quot
(count (word-sequence 1042))
3)
;; => 1

#_(rem
(count (word-sequence 1042))
3)
;; => 1



#_(word-sequence 56945781)
;; => ["50000000" "6000000" "900000" "40000" "5000" "700" "80" "1"]

#_(partition-all 3
(word-sequence 56945781))
;; => (("50000000" "6000000" "900000") ("40000" "5000" "700") ("80" "1"))


#_(quot
(count (word-sequence 56945781))
3)
;; => 2

#_(rem
(count (word-sequence 56945781))
3)
;; => 2


